Patching handling of titles with multiple series

ruoti opened the pull request:
I noticed that there was an issue with the "Collapse Series" option in the Library view that happens when a book is in two series at the same time. For example, in the Pern series of books, the first three books are part of a trilogy (series A) and part of the overall Pern universe (series B). If I label them as such, one of those series will disappear from the view, even though both should be shown. They still show up in the series list, so this is just a display issue.

While fixing this problem, I also noticed that code for sorting books attached to multiple series wouldn't work correctly as sorting happened before collapsing. I've reworked the sorting to have collapsing occur first, then sorting after that. I also fixed a couple of other small issues with sorting:

1. When series are put at the end of the list of titles (happens when not sorting by title), they are now sorted alphabetically.
2. In the above case, a series will appear after books beginning with 'zzzzz', which they wouldn't have before. Admittedly, this would probably never happen, but I was here so I fixed it anyways.
3. If filtering by series, the code would resort after the initial sort. I moved this into the normal sort workflow.

The one non-obvious thing I did was make it so that when filtering by a series and collapsing series at the same time, I do not collapse the titles that are only in the series being filtered. This didn't make sense as you would select a series just to see a single item with that series (unless you had titles with multiple series). Along with my change #3 above, this results in cool functionality where subseries are displayed in order in this case. For example, If there was a trilogy, then a standalone book, then another trilogy, it would show them in this order. Nice for larger series made up of many sub-series (e.g., Dragonlance).

ruoti commented:
I've rebased my commit as there were conflicts with the code pushed last night.

I've also made it so that when collapsing in series view, instead of showing the count for collapsed series, I should the sequence numbers they represent in the filtered series. This matches stylistically with how the sequence number is shown for other books. 

Pre-patch look:
![Screenshot_20221030_114457](https://user-images.githubusercontent.com/2041129/198887924-65c03a54-eb31-4ea0-9531-6083405c08ea.png)

New look:
![Screenshot_20221030_114253](https://user-images.githubusercontent.com/2041129/198887930-495f5491-f9d8-462c-aa39-88222966ba08.png)

Some things to note:
1. With these changes, I think it would make sense to add a "Collapse Series" checkmark to the series view. When dealing with sub-series, I think this works really nicely. Or perhaps, make this just how that view works. It wouldn't have worked in the old scheme where this would have collapsed the main series into itself, but that is no longer the case.
2. I think it is a little weird that the sort order is ignored when filtering by series. For the series tab, this makes sense, but it is a little odd for the others. With the proposed patch, removing this functionality when there are sortBy parameters will work as expected (basing sort by the first box in a collapsed series). However, doing so would make it impossible to sort by sequence numbers in this view, so the drop-down would need to be changed to include that option, though that option wouldn't make sense with other filers. Just something to consider. In this pull request, I've kept the existing functionality the same in this regard.

advplyr commented:
> I've also made it so that when collapsing in series view, instead of showing the count for collapsed series, I should the sequence numbers they represent in the filtered series. This matches stylistically with how the sequence number is shown for other books.

One problem with this is if you have a sub-series with a lot of books then the sequence numbers will overflow.

![image](https://user-images.githubusercontent.com/67830747/199585267-08e362c5-f3a1-4517-ad99-055c2c6eabd9.png)

advplyr commented:
> 2\. I think it is a little weird that the sort order is ignored when filtering by series. For the series tab, this makes sense, but it is a little odd for the others. With the proposed patch, removing this functionality when there are sortBy parameters will work as expected (basing sort by the first box in a collapsed series). However, doing so would make it impossible to sort by sequence numbers in this view, so the drop-down would need to be changed to include that option, though that option wouldn't make sense with other filers. Just something to consider. In this pull request, I've kept the existing functionality the same in this regard.

This is definitely a bug I wasn't aware of. The sort order should not be ignored.

ruoti commented:
Ok, I'm most of the way there. I've fixed the overflow issue, allowed it to display sequence ranges, fixed the item count in the toolbar, and fixed sorting within series view (as this was stated to be a bug). With the latter, it is no longer possible to sort books by sequence in the library view (still works fine in 'book-series' view). To address this, I tried to add another sort field when filtering by series. While I got that working, if I switch to another type of filter, the order by column ends up being blank because the (selected) sequence value is no longer in the list. Considering I've only started using Vue tonight, I'm not expert on how to fix this. I was wondering if you (@advplyr) have any suggestions on how I could reset the order by when no longer filtering by a series.

advplyr commented on server/controllers/LibraryController.js:
```diff
@@ -218,30 +252,55 @@ class LibraryController {
+        // series represents in the filtered series
+        if (filterSeries) {
+          json.collapsedSeries.seriesSequenceList = li.collapsedSeries.books
+            .map(b => parseFloat(b.filterSeriesSequence))
```
An issue with this solution to overflow sequence values is that we don't enforce a sequence to be a number. For example, a valid sequence could be roman numerals or `1a`. In that case this would crash but if we chose to just ignore those then the possibility of the sequence overflowing the book cover would remain.

ruoti replied:
I fixed the logic to use a natural sort. Even without that change, it still wouldn't overflow. The overflow before was caused by a lack of whitespace. I implemented ranges because I think they are more readable, not because they fixed overflow. Below you can see how it looks. I also updated it to use the same styling as normal sequence numbers (I think it looks better).

![Screenshot_20221104_090103](https://user-images.githubusercontent.com/2041129/200104071-3e5927f7-3201-490d-b193-141e087384ce.png)

ruoti commented on client/components/app/BookShelfToolbar.vue:
```diff
@@ -28,13 +28,19 @@
           <div class="flex-grow" />
-          <ui-btn color="primary" small :loading="processingSeries" class="flex items-center" @click="markSeriesFinished">
+          <ui-checkbox v-model="settings.collapseSeries" label="Collapse Series" checkbox-bg="bg" check-color="white"
+            small class="mr-2" @input="updateCollapseSeries" />
```
Updated to use a different setting.

I also added a new commit that moves all of these options to local storage. It took longer than I thought it would, but the utility class I created should help you add any other items you want to move to local storage. Unfortunately, I couldn't get it to work as a mixin, but it isn't too much to add.

ruoti commented:
I believe I have addressed all your comments

advplyr commented:
Thanks for setting up that local storage solution. We will need to set default values for those settings also.
Otherwise we end up with
![image](https://user-images.githubusercontent.com/67830747/200136246-b17587bc-fdb6-4e6a-83e0-3abb1f04c05e.png)

advplyr commented on client/components/app/LazyBookshelf.vue:
```diff
@@ -437,6 +435,9 @@ export default {
         searchParams.set('filter', this.seriesFilterBy)
       } else if (this.page === 'series-books') {
         searchParams.set('filter', `series.${this.$encode(this.seriesId)}`)
+        if (this.collapseSeries) {
```
```suggestion
        if (this.collapseBookSeries) {
```

ruoti commented:
I've rolled back the localStorage changes. I saw that you were doing a fair bit of consistency checking on the server side, and I wasn't ready to go through that logic and make sure it is working in localStorage. I've staged those changes and may go back and add back localStorage in another pull request. This way we don't have to block these changes on localStorage.

ruoti commented:
Also, I probably won't work on the local storage stuff until this is merged so I don't have to worry about rebasing that stuff.

advplyr commented:
Great thanks, working well for me

advplyr commented on server/controllers/LibraryController.js:
```diff
@@ -218,30 +252,55 @@ class LibraryController {
+      libraryItems = naturalSort(libraryItems).by(sortArray)
     }
 
-    if (payload.limit) {
-      var startIndex = payload.page * payload.limit
-      libraryItems = libraryItems.slice(startIndex, startIndex + payload.limit)
-    }
```
Is there a reason this was removed here? Just got notified on Discord about an issue and saw the limit is not applied.

ruoti replied:
No, that was an accident. My guess is that this happened the first time I rebased the patch, where we had both made changes to this file, and I just made a mistake. Sorry.
